[Docs] Refactor and extend structs, unions, enums, function pointer documentation in language basics userguide (#5082)
cython/cython: docs/src/userguide/language_basics.rst

                cdef float *g = [1, 2, 3, 4]
                cdef float *h = &f

.. note::

    There is also support for giving names to types using the
    ``ctypedef`` statement or the ``cython.typedef()`` function, e.g.

    .. tabs::

        .. group-tab:: Pure Python

            .. code-block:: python

                 ULong = cython.typedef(cython.ulong)

                 IntPtr = cython.typedef(cython.p_int)

        .. group-tab:: Cython

            .. code-block:: cython

                ctypedef unsigned long ULong

                ctypedef int* IntPtr

.. _structs:

> Structs, Uions, Enums

Reviewer: ```suggestion
Structs, Unions, Enums
```

---------------------

In addition to the basic types, C :keyword:`struct`, :keyword:`union` and :keyword:`enum`
are supported:

.. tabs::

    .. group-tab:: Pure Python

        .. literalinclude:: ../../examples/userguide/language_basics/struct.py

    .. group-tab:: Cython

        .. literalinclude:: ../../examples/userguide/language_basics/struct.pyx

    .. note::

        Structs can be declared as ``cdef packed struct``, which has

>         the same effect as the C directive ``#pragma pack(1)``.

Reviewer: This doesn't seem to appear in the generated output.

I also wonder if it should have a little more detail on two points:
* removes the empty space between members that C automatically to ensure that they're aligned in memory (just people that don't know what `#pragma pack(1)` does).
* the main use is that numpy structured arrays store their data in packed form, so a `cdef packed struct` can be used in a memoryview to match that.

Author: done

.. tabs::

    .. group-tab:: Pure Python

        .. literalinclude:: ../../examples/userguide/language_basics/union.py

    .. group-tab:: Cython

        .. literalinclude:: ../../examples/userguide/language_basics/union.pyx

.. literalinclude:: ../../examples/userguide/language_basics/enum.pyx

.. note:: Currently, Pure Python mode does not support enums. (GitHub issue :issue:`4252`)

Declaring an enum as ``cpdef`` will create a :pep:`435`-style Python wrapper::

    cpdef enum CheeseState:
        hard = 1
        soft = 2
        runny = 3

There is currently no special syntax for defining a constant, but you can use
an anonymous :keyword:`enum` declaration for this purpose, for example,::

    cdef enum:
        tons_of_spam = 3

.. note::
    In the Cython syntax, the words ``struct``, ``union`` and ``enum`` are used only when
    defining a type, not when referring to it. For example, to declare a variable
    pointing to a ``Grail`` struct, you would write::

        cdef Grail *gp

    and not::

        cdef struct Grail *gp  # WRONG

> You can create a C function by declaring it with :keyword:`cdef` or by decorating a Python function with ``@cfunc``:

Author: Following section/part contains mentions about `cdef` functions and extension types. In general this is wrong place and I think it should be moved:
1. `cdef` function part should be removed because it is better explained in different part (maybe just do double check whether there is not something not mentioned elsewhere.
2. extension type part should be moved somewhere else - not sure where ...

I did not do this changes because I am not 100% sure what is the best. But I am willing to do it after suggestion from others...

Reviewer: I think this PR at least needs an extra heading to separate it from the `structs/unions/enums` section. Not thought about the where it should be long-term

Author: There are already two sections discussing both extension types and C functions:
* C functions - I have removed this part. There is already full pledged section which mentions everything even better
* Extension types - There is another part mentioning that. I have merged them together and created separated heading for that.

                ctypedef int* IntPtr

.. _structs:

Structs, Unions, Enums
----------------------

In addition to the basic types, C :keyword:`struct`, :keyword:`union` and :keyword:`enum`
are supported:

.. tabs::

    .. group-tab:: Pure Python

        .. literalinclude:: ../../examples/userguide/language_basics/struct.py

    .. group-tab:: Cython

        .. literalinclude:: ../../examples/userguide/language_basics/struct.pyx

Structs can be declared as ``cdef packed struct``, which has
the same effect as the C directive ``#pragma pack(1)``:

> .. literalinclude:: ../../examples/userguide/language_basics/packed_struct.pyx

Reviewer: The example is good but I think it's in the wrong place.

I think this might be too complicated for "language basics". I'm wondering if it should be in the "typed memoryviews" part of the documentation with a much shorter bit of text here:

> Structs can be declared as ``cdef packed struct``, which has
the same effect as the C directive ``#pragma pack(1)`` and removes empty space between members. They're most often used for working `[with memoryviews of structures](LINK GOES HERE)`. Pure python mode does not support packed structs.

Sorry for being indecisive. I'm just thinking that this is a bit overwhelming for new users.

Author: I put more complex example since we mention that it is often used with numpy arrays:

> The main use is that numpy structured arrays store their data in packed form

So I thought to put directly example how it can be used with numpy... But I can change it to simple example.

Author: OK. I had a another look. I removed the example left very basic example of definition. I put created there a link to memoryviews. I have left the text as it was before - I like having there link to wikipedia for more explanation about packed structs, etc.

If you want to use these numeric Python types simply omit the
type declaration and let them be objects.

Extension Types
---------------

It is also possible to declare :ref:`extension-types` (declared with ``cdef class`` or the ``@cclass`` decorator).

> Those will have a behaviour very close to python classes (e.g. creating subclasses), but are faster. Typing variable

Reviewer: ```suggestion
Those will have a behaviour very close to python classes (e.g. creating subclasses), but are faster. Typing a variable
```

Reviewer: ```suggestion
Those will have a behaviour very close to python classes (e.g. creating subclasses), but access to their members is faster from Cython code. Typing variable
```

I'm not quite sure what it means for a class to be "faster" so I made it a little more specific

If you want to use these numeric Python types simply omit the
type declaration and let them be objects.

> Extention Types

Reviewer: ```suggestion
Extension Types
```

---------------

It is also possible to declare :ref:`extension-types` (declared with ``cdef class`` or the ``@cclass`` decorator).

> Those will have a behavior very close to python classes (e.g. creating subclasses), but are faster. Typing variable

Reviewer: ```suggestion
Those will have a behaviour very close to python classes (e.g. creating subclasses), but are faster. Typing variable
```

Not very important but I think we've used British/Australian/New Zealand spelling elsewhere

> as extention type is mostly used to access ``cdef``/``@cfunc`` methods and attributes of the extension type.

Reviewer: ```suggestion
as extension type is mostly used to access ``cdef``/``@cfunc`` methods and attributes of the extension type.
```

Function Pointers
-----------------

.. note:: Pointers to functions are currently not supported by pure Python mode. (GitHub issue :issue:`4279`)

> Following example shows declaring ``ptr_add`` function pointer and assigning ``add`` function to it:

Reviewer: ```suggestion
The following example shows declaring a ``ptr_add`` function pointer and assigning the ``add`` function to it:
```
